File: ruby/hamming/hamming.rb
class Hamming

  def self.compute(first, second)
    first, second = first.chars, second.chars
    if first.length < second.length
      shorter, longer = first, second
    else
      shorter, longer = second, first
    end

    shorter.zip(longer).count { |pair| pair[0] != pair[1] }
  end

end

File: javascript/hamming/hamming.js
function countDifferences(reference) {
  return function(count, nucleotide, index) {
    if (nucleotide === reference[index]) {
      return count;
    }
    return count + 1;
  };
}

function compute(strand1, strand2) {
  var shorter, longer;

  if (strand1.length < strand2.length) {
    shorter = strand1;
    longer = strand2;
  } else {
    shorter = strand2;
    longer = strand1;
  }

  return shorter.split('').reduce(countDifferences(longer), 0);
}

if (module) {
  module.exports = {
    compute: compute
  };
}
